In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
#create a model1
model1 =ConcreteModel()

In [ ]:
#we can se certain using python variables
N=14  #we are to declare the number of the variale

In [ ]:
# coefficient for objecetive function can provided using numpy arraysbe 
obj_coef_c=np.array([-4.2, 37.75, 3.5, 10.05,-8.25, 40.85, 0.05, -11.12, 0.65, 20.725, -102.65, 0.6, -9.45, -25.625 ])

In [ ]:
#We use more numpy arrays for providing constraint coefficients
constr1_coef_p =np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
constr2_coef_q=np.array([2.8, -0.6725, 3.6, 0,-2.02, 1.01, -4.305, 0, 1, -2.09, 0, -2.124, -3.402, 0])

#declare the RHS of the constraints as variables
constr1_rhs_b1= 343.48
constr1_rhs_b2= 66.2





In [ ]:
# the following numpy arrays will be useful for setting the lower bound and upper bound for the variable
lower_bound =np.array([2, -3, -np.inf, 1, 7, 3, -9, 0, 1, -25.245, -np.inf, -6, -30.5, -10.8])
upper_bound = np.array([np.inf, 6, 2.5, 3.4, np.inf, 6, 4.1, np.inf, 30, 11, 0, 0, 95, 13.4])

In [ ]:
# set of column indices: since we have N variables, we can take column indices to be the list [0,1,2,...,N-1]
col_indices = np.arange(N)
print(col_indices)
print(col_indices[13])

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13]
13


In [ ]:
#declare the decision variables in the model1
model1.x = Var(col_indices)

In [ ]:
# create a ConstraintList to hold multiple constraints
model1.constraints = ConstraintList()

In [ ]:
#add model constraints one by one to the list 
#first add the constraint p1 x1+ p2 x2+...+ pN xN <= b1 
#model.constr1 = Constraint(expr=4*model.x1 + ....)
model1.constraints.add(sum(constr1_coef_p[j]*model1.x[j] for j in col_indices) <= constr1_rhs_b1)
#np.sum

In [ ]:
#next add the constraint q1 x1 + q2 x2 + ...+ qN xN >= b2 
model1.constraints.add(sum(constr2_coef_q[j]*model1.x[j] for j in col_indices) == constr1_rhs_b2)


In [ ]:
#add the upper bound and lower bounds for the variables
for j in col_indices:
  model1.x[j].setlb(lower_bound[j])
  model1.x[j].setub(upper_bound[j])

In [ ]:
# model1 the objective
model1.objective=Objective(expr=sum(obj_coef_c[j]*model1.x[j] for j in col_indices), sense=minimize)

In [ ]:
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   14 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13}

1 Var Declarations
    x : Size=14, Index=x_index
        Key : Lower   : Value : Upper : Fixed : Stale : Domain
          0 :     2.0 :  None :  None : False :  True :  Reals
          1 :    -3.0 :  None :   6.0 : False :  True :  Reals
          2 :    None :  None :   2.5 : False :  True :  Reals
          3 :     1.0 :  None :   3.4 : False :  True :  Reals
          4 :     7.0 :  None :  None : False :  True :  Reals
          5 :     3.0 :  None :   6.0 : False :  True :  Reals
          6 :    -9.0 :  None :   4.1 : False :  True :  Reals
          7 :     0.0 :  None :  None : False :  True :  Reals
          8 :     1.0 :  No

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 155680 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model1).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -14246.25704375
  Upper bound: -14246.25704375
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 15
  Number of nonzeros: 25
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.05960798263549805
# ----------------------------------------------------------
#   Solution Information
# ---------------

In [ ]:
# display solution
print('\nObjective = ', model1.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model1.x[i].value)

print('\nConstraints')
model1.constraints.display()


Objective =  -14246.257043750004

Decision Variables
x[ 0 ] =  2039.2619375
x[ 1 ] =  -3.0
x[ 2 ] =  -1623.2369375
x[ 3 ] =  1.0
x[ 4 ] =  7.0
x[ 5 ] =  3.0
x[ 6 ] =  -9.0
x[ 7 ] =  0.0
x[ 8 ] =  1.0
x[ 9 ] =  -25.245
x[ 10 ] =  0.0
x[ 11 ] =  -6.0
x[ 12 ] =  -30.5
x[ 13 ] =  -10.8

Constraints
constraints : Size=2
    Key : Lower : Body               : Upper
      1 :  None : 343.47999999999985 : 343.48
      2 :  66.2 :  66.19999999999925 :   66.2


In [ ]:
the optimal value  =  -14246.257043750004


the optimal cost are   =  -14246.257043750004
and the dicision variable are 
 x0 =  2039.2619375
x1  =  -3.0
x2  =  -1623.2369375
x3  =  1.0
x4  =  7.0
x5  =  3.0
x6  =  -9.0
x7  =  0.0
x8 =  1.0
x9 =  -25.245
x10 =  0.0
x11 =  -6.0
x12 =  -30.5
x13 =  -10.8



